#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/RectangleArena.proto"

WorldInfo {
}
Viewpoint {
  orientation 0.0794932508641117 -0.07000376024442917 -0.9943743241952177 4.579977004731579
  position 0.49236771680445124 -2.91506629566836 0.5790393718587727
}
TexturedBackground {
}
TexturedBackgroundLight {
}
RectangleArena {
  floorSize 2 2
}
Robot {
  translation 0 0 0.05
  children [
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0.03 0.035 -0.01
      }
      device [
        RotationalMotor {
          name "Lm_1"
          maxVelocity 40
        }
        PositionSensor {
          name "Lps_1"
        }
      ]
      endPoint Solid {
        translation 0.029999999999999038 0.03500000073106042 -0.009992881850049259
        rotation 1 -1.0850743161034235e-10 1.0842115747089508e-10 1.570000003653369
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColorMap ImageTexture {
                url [
                  ".balance2.jpg"
                ]
              }
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.02
            }
          }
        ]
        name "Lw_1"
        boundingObject USE wheel
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.03 0.035 -0.01
      }
      device [
        RotationalMotor {
          name "Lm_2"
          maxVelocity 40
        }
        PositionSensor {
          name "Lps_2"
        }
      ]
      endPoint Solid {
        translation -0.030000000000000793 0.03500000073106191 -0.009992881850049436
        rotation 1 -1.08500230850091e-10 1.0841395671064371e-10 1.5700000036533575
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColorMap ImageTexture {
                url [
                  ".balance2.jpg"
                ]
              }
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.02
            }
          }
        ]
        name "Lw_2"
        boundingObject USE wheel
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0.03 -0.035 -0.01
      }
      device [
        RotationalMotor {
          name "Rm_1"
          maxVelocity 40
        }
        PositionSensor {
          name "Rps_1"
        }
      ]
      endPoint Solid {
        translation 0.02999999999999998 -0.03499999927332664 -0.009992673248180733
        rotation 1 -1.084118018820952e-10 1.0832559891580637e-10 1.5700000036499653
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColorMap ImageTexture {
                url [
                  ".balance2.jpg"
                ]
              }
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.02
            }
          }
        ]
        name "Rw_1"
        boundingObject USE wheel
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.03 -0.035 -0.01
      }
      device [
        RotationalMotor {
          name "Rm_2"
          maxVelocity 40
        }
        PositionSensor {
          name "Rps_2"
        }
      ]
      endPoint Solid {
        translation -0.030000000000000117 -0.034999999273325255 -0.009992673248180726
        rotation 1 -1.0841550779483571e-10 1.0832931219128741e-10 1.5700000036499646
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColorMap ImageTexture {
                url [
                  ".balance2.jpg"
                ]
              }
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.01
              radius 0.02
            }
          }
        ]
        name "Rw_2"
        boundingObject USE wheel
        physics Physics {
        }
      }
    }
    DEF Body Pose {
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 1 0.6 1
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.1 0.06 0.03
          }
        }
      ]
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
      }
      device [
        PositionSensor {
          name "rod_ps"
        }
      ]
      endPoint Solid {
        translation -5.421010862427522e-20 0 0.05999999999999997
        rotation -5.6750389742651486e-05 -0.0031165319064336157 0.9999951419923345 6.832405839897413e-06
        children [
          DEF rod Shape {
            appearance PBRAppearance {
              baseColor 1 1 0.5
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.12
              radius 0.002
            }
          }
        ]
        name "pendulum"
        boundingObject USE rod
        physics Physics {
          mass 0.2
        }
      }
    }
  ]
  boundingObject USE Body
  physics Physics {
    mass 2
    centerOfMass [
      0 0 0
    ]
  }
  controller "inv_pendulum_ctrl"
}
